## ML5Assign: Nashville Housing - Machine Learning 



In [1]:
! git clone https://github.com/gacree/NashvilleHousing.git

Cloning into 'NashvilleHousing'...
remote: Enumerating objects: 9, done.
remote: Counting objects: 100% (9/9), done.
remote: Compressing objects: 100% (7/7), done.
remote: Total 9 (delta 2), reused 0 (delta 0), pack-reused 0
Receiving objects: 100% (9/9), 7.58 MiB | 3.38 MiB/s, done.
Resolving deltas: 100% (2/2), done.


Import packages, read data into dataframes, reassign column names



In [2]:
#import the pandas package for data analytics.
import pandas as pd

# Read the data files into pandas dataframes.
# Notice that we must indicate the full path of the data file.
houses_2017 = pd.read_csv('NashvilleHousing/2017SINGLEFAMILYSF.txt')

# Alter the column names in each of the new dataframes.
houses_2017.columns = ['APN', 'DistrictCode', 'CouncilDistrict', 'AddressFullAddress',
       'AddressCity', 'AddressPostalCode', 'LAND', 'IMPR', 'TOTALAPPR',
       'TOTALASSD', 'FinishedArea']

### Create a new column 'CityNum' that will identify each 'AddressCity' with an integer value from 0-15.

In [3]:
# List the unique AddressCity values
# Note there are 16 of them.
houses_2017['AddressCity'].unique()

array(['NASHVILLE', 'GOODLETTSVILLE', 'JOELTON', 'WHITES CREEK',
       'MADISON', 'OLD HICKORY', 'ASHLAND CITY', 'HERMITAGE', 'PEGRAM',
       'MOUNT JULIET', 'FRANKLIN', 'ANTIOCH', 'BELLEVUE', 'BRENTWOOD',
       'LAVERGNE', 'NOLENSVILLE'], dtype=object)

In [6]:
# df[newcolumn] = pd.factorize(df[column])[0]
# This method will take all the unique values of df[column]
# and change them to integers from 0 - len(column) 

# pd.factorize() has two outputs ... 
# The first gives the integers from 0 - len(column) and
# The second gives the original unique values of the column
# Since we're interested in these first values we append [0]

houses_2017['CityNum'] = pd.factorize(houses_2017['AddressCity'])[0]

houses_2017[['AddressCity','CityNum']].sample(20)

AddressCity  CityNum
131965       NASHVILLE        0
19499          MADISON        4
192902         ANTIOCH       11
191936         ANTIOCH       11
146018       NASHVILLE        0
28820        NASHVILLE        0
161387         ANTIOCH       11
29746        NASHVILLE        0
89034        NASHVILLE        0
151003       NASHVILLE        0
65004        HERMITAGE        7
1126    GOODLETTSVILLE        1
163952         ANTIOCH       11
70525        NASHVILLE        0
94694        NASHVILLE        0
36073        HERMITAGE        7
180691         ANTIOCH       11
181377         ANTIOCH       11
48341        HERMITAGE        7
180315         ANTIOCH       11

In [27]:
houses_2017['AddressCity'].value_counts()

NASHVILLE         134723
ANTIOCH            22770
HERMITAGE          10894
MADISON             9771
OLD HICKORY         5380
BRENTWOOD           4983
GOODLETTSVILLE      4422
JOELTON             1506
WHITES CREEK        1238
NOLENSVILLE          800
MOUNT JULIET         266
PEGRAM               173
ASHLAND CITY         157
FRANKLIN              26
LAVERGNE               3
BELLEVUE               2
Name: AddressCity, dtype: int64

In [29]:
houses_2017.sample(10)

APN DistrictCode  CouncilDistrict    AddressFullAddress  \
131617     13210004100         OH                 25     805 CURTISWOOD LN   
187565  173040A27200CO         USD                31     5449 OAK CHASE DR   
121858  126160A47400CO         GSD                35    7131 RIVERFRONT DR   
40900      07202002000         USD                 8      1013 IVERSON AVE   
88976      10304027100         USD                24        4101 IDAHO AVE   
27176      05914020700         USD                 2        3830 ATWELL DR   
131483     13209009200         OH                 25     4410 CADILLAC AVE   
44840   072150A10000CO         USD                 7        100 PORTER TER   
100832     10705018200         USD                13        1901 ELANOR DR   
157393  149050A23800CO         USD                28  238 PEPPER RIDGE CIR   

       AddressCity  AddressPostalCode    LAND     IMPR  TOTALAPPR  TOTALASSD  \
131617   NASHVILLE              37204  920000  3249000    4169000    1042250   
187565     ANTIOCH              37013   41500   266700     308200      77050   
121858   NASHVILLE              37221   54000   215500     269500      67375   
40900    NASHVILLE              37216   85000   214700     299700      74925   
88976    NASHVILLE              37209  225000   378200     603200     150800   
27176    NASHVILLE              37207   29000   120600     149600      37400   
131483   NASHVILLE              37204  361800    57200     419000     104750   
44840    NASHVILLE              37206   46000   127200     173200      69280   
100832   NASHVILLE              37217   21500    69200      90700      22675   
157393     ANTIOCH              37013   24000    94300     118300      29575   

        FinishedArea  CityNum  
131617   11747.41992        0  
187565    3337.80005       11  
121858    2254.00000        0  
40900     1273.00000        0  
88976     3840.00000        0  
27176     1102.00000        0  
131483       0.00000        0  
44840      896.00000        0  
100832     975.00000        0  
157393    1341.00000       11

# Your objective in this notebook is to increase the Model.score(X_test, y_test)) based upon the ML model predictions of the CityNum based upon the quantitative features known about each property.

### You may alter the choice of data features (or even the target values).
X = houses_2017[['LAND', 'IMPR', 'TOTALAPPR', 'TOTALASSD', 'FinishedArea']]

y = houses_2017['CityNum']
### -

### You may also choose to alter the test_size value when you split the data into training and testing datasets.
test_size= ?

### -

### In limited ways you can also 'play' with the model parameters such as ...

n_neighbors= ? (Nearest Neighbors)

n_estimators= ? (Random Forest)


### 1) In the following cells make changes to your choices of features included in X. Then run Models One & Four below to determine whether or not your predictions have improved.
Note: This may take several iterations before you are content with the final results.



In [30]:
##### Don't Change This Cell #####
##### In fact, you don't even need to execute this cell.
##### If you make changes to the cell below ... 
##### that causes the cell to no longer function ... 
##### you can always copy the following 3 lines below and begin fresh!

# Define the features and the target from our dataset
X = houses_2017[['LAND', 'IMPR', 'TOTALAPPR', 'TOTALASSD', 'FinishedArea']]
y = houses_2017['CityNum']

In [31]:
# ???
##### Make your changes to X and y in this cell!!
##### You have many options:
##### Try one feature at a time, or two, or three, or four, or all
##### In any combination.
##### You certainly don't need to try them all, but give it some thought
##### And determine what you think might work best.

# Define the features and the target from our dataset ... for instance:
X = houses_2017[['LAND',  'IMPR',  'FinishedArea', 'TOTALASSD', 'TOTALAPPR']]
y = houses_2017['CityNum']

### Create Training and Testing Sets

In [36]:
# import the package to create the Training and Testing split.
from sklearn.model_selection import train_test_split

# The code below splits our X and y into training and testing sets. 
# Since test_size=0.30, the split will be 70% Training to 30% Testing sets.
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=2)

## Model One: Select, Instantiate, Fit, Score

In [56]:
##### Don't make changes to this cell ...
##### Just execute it ...
##### After each adjustment to your feature selection

## Select/Import the Model.
from sklearn.neighbors import KNeighborsClassifier
    
## Instantiate the Knn model with ... neighbors. 
ModelOne = KNeighborsClassifier(n_neighbors=3)
    
## Fit the model on the training data.
ModelOne.fit(X_train, y_train)
    
## See how the model performs on the test data.
print(ModelOne.score(X_test, y_test))

0.7549907414453492


## Model Four: Select, Instantiate, Fit, Score

In [58]:
##### Don't make changes to this cell ...
##### Just execute it ...
##### After each adjustment to your feature selection

## Select/Import the Model.
from sklearn.ensemble import RandomForestClassifier

## Instantiate the Random Forest model
ModelFour = RandomForestClassifier(n_estimators=10)

## Fit the model on the training data.
ModelFour.fit(X_train, y_train)

## See how the model performs on the test data.
print(ModelFour.score(X_test, y_test))

0.8086649925170586


### 2) After you have determined which feature selection gives you the best scores ...

### In the text cell below, briefly describe the changes you made to the features, how these changes improved your results, and possible reasons for this improvement.

I increased the n_estimators for ModelFour to 15. It got better as I increased it but it started taking long processing time. ModelOne gave the best performance at n_neighbors = 3.

### 3) Try other values for test_size in the following code. Then run the models below to determine whether or not your predictions have improved.

Note: Again, this may take several iterations before you are content with the final results.

In [73]:
# ???
# import the package to create the Training and Testing split.
from sklearn.model_selection import train_test_split

##### Change the current value of test_size ... until you are satisfied ...
##### That you cannot make additional improvement in the scores to follow.
##### Note: It must be between 0.0 and 1.0 ... 
##### But these extremes will not be very helpful!!!

# The code below splits our X and y into training and testing sets. 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=2)

## Model One: Select, Instantiate, Fit, Score

In [68]:
##### Don't make changes to this cell ...
##### Just execute it ...
##### After each adjustment to your feature selection

## Select/Import the Model.
from sklearn.neighbors import KNeighborsClassifier
    
## Instantiate the Knn model with ... neighbors. 
ModelOne = KNeighborsClassifier(n_neighbors=3)
    
## Fit the model on the training data.
ModelOne.fit(X_train, y_train)
    
## See how the model performs on the test data.
print(ModelOne.score(X_test, y_test))

0.7549907414453492


## Model Four: Select, Instantiate, Fit, Score

In [74]:
##### Don't make changes to this cell ...
##### Just execute it ...
##### After each adjustment to your feature selection

## Select/Import the Model.
from sklearn.ensemble import RandomForestClassifier

## Instantiate the Random Forest model
ModelFour = RandomForestClassifier(n_estimators=15)

## Fit the model on the training data.
ModelFour.fit(X_train, y_train)

## See how the model performs on the test data.
print(ModelFour.score(X_test, y_test))

0.813636709535043


### 4) In the following cell briefly describe the changes you made above, how they improved your results, and possible reasons for this improvement.

I changed the train test split to 80-20 ratio which improved the results. It might be because there was more data to train on. Or it could be because model overfit for the smaller test data.

### 5) Try other values for n_neighbors in the following code. Then run the model below to determine whether or not your predictions have improved.

Note: As a simplification ... the K Nearest Neighbors Model predicts the target of a test element by looking at the targets of the K elements in the training set that are 'closest' to it. Changing K (the number of neighbors) can change your results.

## Model One: Select, Instantiate, Fit, Score

In [82]:
# ???
##### Note: The only change you need to make is to the number for n_neighbors= ?
##### Change the current value of n_neighbors ... until you are satisfied ...
##### That you cannot make additional improvement in the scores to follow.

## Select/Import the Model.
from sklearn.neighbors import KNeighborsClassifier
    
## Instantiate the Knn model with ... neighbors. 
ModelOne = KNeighborsClassifier(n_neighbors=3)
    
## Fit the model on the training data.
ModelOne.fit(X_train, y_train)
    
## See how the model performs on the test data.
print(ModelOne.score(X_test, y_test))


0.7549907414453492


### 6) In the following cell briefly describe the changes you made above, how they improved your results, and possible reasons for this improvement.

I changed the n_neighbors to different values but n=3 gave the best results.

### 7) Try other values for n_estimators in the following code. Then run the model below to determine whether or not your predictions have improved. 

Note: As a simplification ... the Random Forest Model uses multiple Decision Trees (of course) and chooses the best among them. Changing the number of Decision Trees considered could alter your results.

## Model Four: Select, Instantiate, Fit, Score

In [83]:
# ???
##### Note: The only change you need to make is to the number for n_estimators= ?
##### Change the current value of n_estimators= ... until you are satisfied ...
##### That you cannot make additional improvement in the scores to follow.

## Select/Import the Model.
from sklearn.ensemble import RandomForestClassifier

## Instantiate the Random Forest model
ModelFour = RandomForestClassifier(n_estimators=15)

## Fit the model on the training data.
ModelFour.fit(X_train, y_train)

## See how the model performs on the test data.
print(ModelFour.score(X_test, y_test))

0.8139918321791847


### 8) In the following cell briefly describe the changes you made above, how they improved your results, and possible reasons for this improvement.

Changing n_estimators to higher numbers gave better results with more processing time.

### 9) What was your overall best score? ... and from which model did it come?

It was 0.8139918321791847 with RandomForestClassifier.

# Optionally for those looking for a bit of a challenge:

### 10) Take the time to redefine the values in 'CityNum' to reduce the number of possible targets. Then run the models below to determine whether or not your predictions have improved.

Note1: There are several options here: You might limit your dataframe to only a few of the AddressCity values (rather than the full 16) or you might have more than one AddressCity value lead to a single CityNum value ... grouping some of the AddressCity values together. You might think of other possibilities as well.

Note2: Again, this may take several iterations before you are content with the final results.

In [ ]:
# This cell is only included for a reference if needed. 
#Use, alter, or ignore at your own discretion.

# Create a new column and give it an initial value.
houses_2017['CityNum'] = 99

# Create a counter that will be incremented to assign distinct values to each AddressCity
i = 0

# Loop through each of the unique AddressCity values 
for city in houses_2017['AddressCity'].unique():
  # assign the counter value to the 'CityNum' for that AddressCity
  houses_2017.loc[houses_2017['AddressCity']== city, 'CityNum'] = i
  # increment the counter for use for the next AddressCity
  i+=1
  
# View a sample of the resulting dataframe
houses_2017[['AddressCity','CityNum']].sample(20)

# Notice that the assigned numbers are given 
# in the order of houses_2017['AddressCity'].unique() as seen above.
# NASHVILLE - 0, GOODLETTSVILLE - 1, JOELTON - 2, ...

In [78]:
# ???
# Redefine the values in 'CityNum'

# Change all the values of 'CityNum' to a specific number.
houses_2017['CityNum'] = 99

# Hint: Rather than looping through all the 'Addresscity' cities, ...
# you might use something of the following form (for each value you wish to change):
houses_2017.loc[houses_2017['AddressCity']== 'NASHVILLE', 'CityNum'] = 0
houses_2017.loc[houses_2017['AddressCity']== 'JOELTON', 'CityNum'] = 1
# Note: So if you make no changes in this cell then the value in 'CityNum' ...
# Will be 0 if 'AddressCity' is 'NASHVILLE', 
# 1 if 'AddressCity' is 'JOELTON' and 99 otherwise.
# Thus the models are only predicting if a housing unit is in NASHVILLE,
# JOELTON,  or not.
# What other cities (or groups of cities) might you choose to consider?
# Which are easier to predict? Which are more difficult?


### Create Training and Testing Sets

In [79]:
# import the package to create the Training and Testing split.
from sklearn.model_selection import train_test_split

# The code below splits our X and y into training and testing sets. 
# Since test_size=0.30, the split will be 70% Training to 30% Testing sets.
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.10, random_state=2)

## Model One: Select, Instantiate, Fit, Score

In [80]:
##### Don't make changes to this cell ... except possibly to n_neighbors.
##### Just execute it ...
##### After each adjustment to your feature selection

## Select/Import the Model.
from sklearn.neighbors import KNeighborsClassifier
    
## Instantiate the Knn model with ... neighbors. 
ModelOne = KNeighborsClassifier(n_neighbors=3)
    
## Fit the model on the training data.
ModelOne.fit(X_train, y_train)
    
## See how the model performs on the test data.
print(ModelOne.score(X_test, y_test))

0.8167106331168831


In [81]:
houses_2017[['AddressCity','CityNum']].sample(30)

AddressCity  CityNum
116220    NASHVILLE        0
157581      ANTIOCH       99
25727     NASHVILLE        0
109981    NASHVILLE        0
129488    NASHVILLE        0
137652    NASHVILLE        0
144259    NASHVILLE        0
40179     NASHVILLE        0
186812    NASHVILLE        0
71300     NASHVILLE        0
114670    NASHVILLE        0
194921    BRENTWOOD       99
63091     NASHVILLE        0
64119     NASHVILLE        0
49996     HERMITAGE       99
155728    NASHVILLE        0
130814    NASHVILLE        0
18887       MADISON       99
67150     HERMITAGE       99
87438     NASHVILLE        0
27775     NASHVILLE        0
172443    NASHVILLE        0
42710     NASHVILLE        0
162889      ANTIOCH       99
191433      ANTIOCH       99
51798     NASHVILLE        0
23094   OLD HICKORY       99
189691      ANTIOCH       99
98894     NASHVILLE        0
154056      ANTIOCH       99

## Model Four: Select, Instantiate, Fit, Score

In [76]:
##### Don't make changes to this cell ... except possibly to n_estimators.
##### Just execute it ...
##### After each adjustment to your feature selection

## Select/Import the Model.
from sklearn.ensemble import RandomForestClassifier

## Instantiate the Random Forest model
ModelFour = RandomForestClassifier(n_estimators=3)

## Fit the model on the training data.
ModelFour.fit(X_train, y_train)

## See how the model performs on the test data.
print(ModelFour.score(X_test, y_test))

0.9912743506493507


### 11) In the following cell briefly describe the changes you made above, how they improved your results, and possible reasons for this improvement.

'# ???'